statistics/statistics.go

```diff
@@ -280,56 +280,129 @@ func (p *Plugin) GetConfigPolicy() (*cpolicy.ConfigPolicy, error) {
 	cp.Add([]string{""}, config)
 
 	return cp, nil
+
+}
+
+type byTimestamp []plugin.MetricType
+
+func (sa byTimestamp) Len() int {
+	return len(sa)
+}
+
+func (sa byTimestamp) Less(i, j int) bool {
+	return sa[i].Timestamp().Before(sa[j].Timestamp())
+
+}
+func (sa byTimestamp) Swap(i, j int) {
+	sa[i], sa[j] = sa[j], sa[i]
+}
+
+func (p *Plugin) insertInToTimeBuffer(metric plugin.MetricType, times []time.Time) []time.Time {
+	if p.bufferCurSize == 0 {
+		times[0] = metric.Timestamp()
+	} else {
+		times[p.bufferIndex] = metric.Timestamp()
+	}
+	return times
+}
+
+func (p *Plugin) getTimes(times []time.Time) (time.Time, time.Time) {
+	//log.Printf("Buffer index: %v", metric.Timestamp())
+	if p.bufferCurSize == p.bufferMaxSize && p.bufferIndex != p.bufferMaxSize-1 {
+		return times[p.bufferIndex+1], times[p.bufferIndex]
+	}
+	return times[0], times[p.bufferIndex]
 }
 
 // Process processes the data, inputs the data into this' buffer and calls the descriptive statistics method
 func (p *Plugin) Process(contentType string, content []byte, config map[string]ctypes.ConfigValue) (string, []byte, error) {
 	var metrics []plugin.MetricType
-
+	p.bufferMaxSize = 100
 	if config != nil {
 		if config["SlidingWindowLength"].(ctypes.ConfigValueInt).Value > 0 {
 			p.bufferMaxSize = config["SlidingWindowLength"].(ctypes.ConfigValueInt).Value
-		} else {
-			p.bufferMaxSize = 100
 		}
-	} else {
-		p.bufferMaxSize = 100
 	}
-
 	//Decodes the content into PluginMetricType
 	dec := gob.NewDecoder(bytes.NewBuffer(content))
 	if err := dec.Decode(&metrics); err != nil {
 		log.Printf("Error decoding: error=%v content=%v", err, content)
 		return "", nil, err
 	}
+	var results []plugin.MetricType
+	times := make([]time.Time, p.bufferMaxSize)
 
+	metricNamespace := make(map[string][]plugin.MetricType)
 	for _, metric := range metrics {
-		log.Printf("Data received %v", metric.Data())
-		p.insertInToBuffer(metric.Data(), metric.Namespace().Strings())
+		//Populates the metricNamespace map so that the statistics are ran on metrics that share the same namespace.
+		ns := concatNameSpace(metric.Namespace().Strings())
+		if plugins, ok := metricNamespace[ns]; ok {
+			plugins = append(plugins, metric)
+			metricNamespace[ns] = plugins
+		} else {
+			metricNamespace[ns] = []plugin.MetricType{metric}
+		}
 	}
 
-	p.updateCounters()
+	for k, v := range metricNamespace {
+		var startTime time.Time
+		var stopTime time.Time
+		unit := v[0].Unit()
+
+		sort.Sort(byTimestamp(v))
+		for _, metric := range v {
+			switch reflect.ValueOf(metric.Data()).Kind() {
```

Comment:
For each individual type, there is really no operation instead of insertion using interface. You can remove switch statement totally. This should do it: p.insertInToBuffer(metric.Data(), metric.Namespace().Strings())